pycalphad/model.py

```diff
@@ -422,15 +422,14 @@ def build_phase(self, dbe):
     def _purity_test(self, constituent_array):
         """
         Check if constituent array only has one species in its array
-        This species must also be an active species
+        This species must also be an active species and contained in model sublattice
         """
-        if len(constituent_array) != len(self.constituents):
+        if not self._array_validity(constituent_array):
             return False
-        for sublattice in constituent_array:
-            if len(sublattice) != 1:
+        for param_sublattice, model_sublattice in zip(constituent_array, self.constituents):
```

Comment:
Correct me if I’m wrong, but shouldn’t the `_array_validity` check assure that `param_sublattice[0]` is always in the model sublattice or is a `*` (making the check redundant)? This seems like it could break in the case of a `*` anyway. 

If that’s true, could we just change the lines after the validity check to 

```python
return any(len(sublattice) != 1 for sublattice in constituent_array)
```

Reply:
`_array_validity` really checks if `param_sublattice[0]` is always in the model sublattice or is a `*`, but it does not ensure that `param_sublattice[0]` is contained in the correspondent Model sublattice. Actually, `if param_sublattice[0] not in model_sublattice` is the code line that fixes the bug related in the issue 256.
`_array_validity` returns True if each sublattice species is contained at any Model sublattice, while these changes ensure that each parameter sublattice is contained in respective Model sublattice (respective sublattice, instead of any sublattice).
Now seeing this, I realized: does not make more sense to perform the checking if parameter sublattices are contained in respective Model sublattices in `_array_validity` and just check if there are one specie by sublattice in `_purity_test`?

Reply:
Yes, you’re right. I thought that’s what `_array_validity` was doing, perhaps I shouldn’t review code so late at night 😬

> Now seeing this, I realized: does not make more sense to perform the checking if parameter sublattices are contained in respective Model sublattices in `_array_validity` and just check if there are one specie by sublattice in `_purity_test`?

Yeah, this makes more sense to me, since that requirement is common to both. I think the only thing that needs to change in `_array_validity` is to check if the sublattice constituents are a subset of the Model sublattice, rather than the components. 

If we do that, then we can make the simplifications I suggest in these comments.

Reply:
Not problem about reviewing code late at night! I think that I applied all suggestions in new sent commits!